PolyEngine/Core/Src/pe/core/utils/Result.hpp

```diff
@@ -0,0 +1,117 @@
+#pragma once
+
+#include <pe/Defines.hpp>
+#include <variant>
+
+namespace pe::core::utils
+{
+	// 
+	template <typename T, typename E>
+	class Result
+	{
+	public:
+		Result() = delete;
+		Result(Result&) = default;
+		Result(Result&&) = default;
+		Result& operator=(Result&) = default;
+		Result& operator=(Result&&) = default;
+		Result(T okVal) : m_value(std::move(okVal)) {}
+		Result(E errVal) : m_value(std::move(errVal)) {}
+		
+		bool isOk() const { return m_value.index() == 0; }
+		bool isErr() const { return !isOk(); }
+		
+		T& getValue() { return std::get<T>(m_value); }
+		E& getError() { return std::get<E>(m_value); }
+		const T& getValue() const { return std::get<T>(m_value); }
+		const E& getError() const { return std::get<E>(m_value); }
+
+		template <typename U>
+		Result<U, E> and(Result<U, E> res)
+		{
+			if (isOk())
+				return res;
+			else
+				return getError();
+		}
+
+		template <typename F>
+		Result<T, F> or(Result<T, F> res)
+		{
+			if (isOk())
+				return getValue();
+			else
+				return res;
+		}
+
+	private:
+		std::variant<T, E> m_value;
+	};
+	
+	template <typename E>
+	class Result<void, E> final : public BaseObjectLiteralType<>
+	{
+	public:
+		Result() = default;
+		Result(Result&) = default;
+		Result(Result&&) = default;
+		Result& operator=(Result&) = default;
+		Result& operator=(Result&&) = default;
+		Result(E errVal) : m_value(std::move(errVal)) {}
+
+		bool isOk() const { return m_value.has_value() == false; }
+		bool isErr() const { return !isOk(); }
+
+		E& getError() { return m_value.value(); }
+		const E& getError() const { return m_value.value(); }
+
+		template <typename U>
+		Result<U, E> and(Result<U, E> res)
+		{
+			if (isOk())
+				return res;
+			else
+				return getError();
+		}
+
+		template <typename F>
+		Result<void, F> or (Result<void, F> res)
+		{
+			if (isOk())
+				return Result<void, F>();
+			else
+				return res;
+		}
```

Comment:
We need to check the behavior of those functions in Rust, as current implementation is strange.